src/zenml/materializers/built_in_materializer.py

```diff
@@ -66,11 +71,246 @@ def handle_input(self, data_type: Type[Any]) -> Any:
+                element = materializer.handle_input(type_)
+                outputs.append(element)
+
+        # Cast the data to the correct type.
+        if issubclass(data_type, dict) and not isinstance(outputs, dict):
+            keys, values = outputs
+            return dict(zip(keys, values))
+        if issubclass(data_type, tuple):
+            return tuple(outputs)
+        if issubclass(data_type, set):
+            return set(outputs)
+        return outputs
+
+    def handle_return(self, data: Any) -> None:
+        """Materialize a built-in container object.
+
+        If the object can be serialized to JSON, serialize it.
+
+        Otherwise, use the `default_materializer_registry` to find the correct
+        materializer for each element and materialize each element into a
+        subdirectory.
+
+        Tuples and sets are cast to list before materialization.
+
+        For non-serializable dicts, materialize keys/values as separate lists.
+
+        Args:
+            data: The built-in container object to materialize.
+
+        Raises:
+            Exception: If any exception occurs, it is raised after cleanup.
+        """
+        super().handle_return(data)
+
+        # tuple and set: handle as list.
+        if isinstance(data, tuple) or isinstance(data, set):
+            data = list(data)
+
+        # If the data is serializable, just write it into a single JSON file.
+        if _is_serializable(data):
+            yaml_utils.write_json(self.data_path, data)
+
+        # non-serializable dict: Handle as non-serializable list of lists.
+        if isinstance(data, dict):
+            data = [list(data.keys()), list(data.values())]
+
+        # non-serializable list: Materialize each element into a subfolder.
+        # Get path, type, and corresponding materializer for each element.
+        paths, types, materializers = [], [], []
+        for i, element in enumerate(data):
+            element_path = os.path.join(self.artifact.uri, str(i))
+            os.mkdir(element_path)
+            type_ = type(element)
+            paths.append(element_path)
+            types.append(str(type_))
+            materializer_class = default_materializer_registry[type_]
+            mock_artifact = DataArtifact()
+            mock_artifact.uri = element_path
+            materializer = materializer_class(mock_artifact)
+            materializers.append(materializer)
+        try:
+            # Write metadata as JSON.
+            metadata = {"length": len(data), "paths": paths, "types": types}
+            yaml_utils.write_json(self.metadata_path, metadata)
+            # Materialize each element.
+            for element, materializer in zip(data, materializers):
+                materializer.handle_return(element)
+        # If an error occurs, delete all created files.
+        except Exception as e:
+            # Delete metadata
+            if os.path.exists(self.metadata_path):
```

Comment:
Fileio here, in the `remove` line and the `rmtree`

Reply:
also for the `exists`?